ext/MitsubaLoader/CMitsubaLoader.h

```diff
@@ -16,255 +20,50 @@ namespace ext
-		//2 weights in order to encode MIXTURE bsdf as multiple BLENDs
-		//if flag decides to use weight texture, `weightL` is bindless texture ID and weightR is then irrelevant
-		//otherwise `weightL` and `weightR` are constant RGB19E7 blend weights. Left has to be multiplied by weightL and right operand has to be multiplied by weightR.
-		uint64_t weightL;
-		uint64_t weightR;
-		padding_t<2, max_bsdf_struct_size> padding;
-	} PACK_STRUCT;
-#include "irr/irrunpack.h"
-
-	union SBSDFUnion
-	{
-		SAllDiffuse diffuse;
-		SDiffuseTransmitter diffuseTransmitter;
-		SAllDielectric dielectric;
-		SAllConductor conductor;
-		SAllPlastic plastic;
-		SAllCoating coating;
-		SBumpMap bumpmap;
-		SWard ward;
-		SMixture mixture;
-		SBlend blend;
-	};
-}
-
 class CElementBSDF;
-
-struct NastyTemporaryBitfield
-{
-#define MITS_TWO_SIDED		0x80000000u
-#define MITS_USE_TEXTURE	0x40000000u
-#define MITS_BUMPMAP		0x20000000u
-	uint32_t _bitfield;
-};
+class CMitsubaMaterialCompilerFrontend;
 
 class CMitsubaLoader : public asset::IAssetLoader
 {
+		friend class CMitsubaMaterialCompilerFrontend;
 	public:
 		//! Constructor
 		CMitsubaLoader(asset::IAssetManager* _manager);
 
 	protected:
-		asset::IAssetManager* manager;
+		asset::IAssetManager* m_manager;
 
-		struct SContext
+#include "irr/irrpack.h"
+		//compatible with std430 and std140
+		struct SInstanceData
 		{
-			const asset::IGeometryCreator* creator;
-			const asset::IMeshManipulator* manipulator;
-			const asset::IAssetLoader::SAssetLoadParams params;
-			asset::IAssetLoader::IAssetLoaderOverride* override;
-			CGlobalMitsubaMetadata* globalMeta;
-
-			//
-			using group_ass_type = core::smart_refctd_ptr<asset::ICPUMesh>;
-			core::map<const CElementShape::ShapeGroup*, group_ass_type> groupCache;
-			//
-			using shape_ass_type = core::smart_refctd_ptr<asset::ICPUMesh>;
-			core::map<const CElementShape*, shape_ass_type> shapeCache;
-			//! TODO: change to CPU graphics pipeline
-			using bsdf_ass_type = core::smart_refctd_ptr<asset::ICPURenderpassIndependentPipeline>;
-			core::map<const CElementBSDF*, bsdf_ass_type> pipelineCache;
-			//! TODO: even later when texture changes come, might have to return not only a combined sampler but some GLSL sampling code due to the "scale" and offset XML nodes
-			using tex_ass_type = std::pair<core::smart_refctd_ptr<asset::ICPUImageView>,core::smart_refctd_ptr<asset::ICPUSampler> >;
-			core::unordered_map<const CElementTexture*, tex_ass_type> textureCache;
-
-			core::vector<bsdf::SBSDFUnion> bsdfBuffer;
-			core::vector<bsdf::instr_t> instrBuffer;
-			uint32_t instrBufOffset;
-		};
+			core::matrix3x4SIMD tform;//mat4x3
+			//elements (0,3) and (1,3) are (first,count) for bsdf_instrStream (remainder and pdf stream in case of RT backend)
+			core::matrix3x4SIMD normalMatrix;
+			std::pair<uint32_t,uint32_t> prefetch_instrStream;
+			std::pair<uint32_t, uint32_t> nprecomp_instrStream;
+			std::pair<uint32_t, uint32_t> genchoice_instrStream;
```

Comment:
I thought we could have these as 4 values, 1 offset and 3 counts ?

Reply:
yes, never actually got to it though. A minor TODO